### PR TITLE
Sentinel 2 NBART S3 Sync

### PR DESCRIPTION
**Pull request by kieranricardo**
This PR adds a DAG + uploader script that syncs S2 NBART data from NCI to S3. The DAG is almost identical to the [nbar dag](https://github.com/GeoscienceAustralia/dea-airflow/blob/master/dags/nci_s2_upload_s3.py), and the upload script has a few additions changes.

The upload script now:

- creates eo3 and stac metadata and uploads this to S3
- sends the stac metadata to the SNS topic "U29500-Test"

### CODE REVIEW

**Comment by kieranricardo** [comment]
@alexgleith @pindge sorry don't review yet - I'm using a later version of airflow locally and need to fix a few things up

**Comment by kieranricardo** [comment]
alrighty, ready for review :)

File: dags/sentinel_2_nbart/nbart_nci_s2_upload_s3.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    # Execute script to upload sentinel-2 data to s3 bucket
+    aws_hook = AwsHook(aws_conn_id=dag.default_args['aws_conn_id'])
+
+    execute_upload = SSHOperator(
+        task_id='execute_upload',
+        # language="Shell Script"
+        command=dedent(COMMON + """
+            {% set aws_creds = params.aws_hook.get_credentials() -%}
+            # Export AWS Access key/secret from Airflow connection module
+            export AWS_ACCESS_KEY_ID={{aws_creds.access_key}}
```
Comment (Kirill888) [inline dags/sentinel_2_nbart/nbart_nci_s2_upload_s3.py:108]: Does this end up on disk or is this just running inside ssh process?
If on-disk we better make sure files are not world readable.

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+
+def setup_logging():
+    """Log to stdout (via TQDM if running interactively) as well as into a file."""
+    _LOG.setLevel(logging.INFO)
+    if sys.stdout.isatty():
+        c_handler = TqdmLoggingHandler()
+    else:
+        c_handler = logging.StreamHandler()
+    f_handler = logging.FileHandler('s3_uploads.log')
```
Comment (Kirill888) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:51]: need to make sure that logs don't have `AWS_*` things in there. 

Best also to make sure at higher level that new files are not group:other readable.

File: dags/sentinel_2_nbart/upload_s2_nbart.py [hunk 3]
```diff
@@ -95,16 +96,39 @@ def main(granule_ids, workers):
...
+    """
+    Checks to see whether the folder `granule_id` exists
+
+    :param granule_id: the id of the granule in format 'date/tile_id'
+    :param session: boto3 Session object
+    :return: bool - `True` indicates that ganule has been uploaded
+    """
+    s3 = session.resource('s3')
+    try:
+        _ = s3.Object(S3_BUCKET, f"{S3_PATH}/{granule_id}/stac-ARD-METADATA.json").load()
```
Comment (Kirill888) [inline dags/sentinel_2_nbart/upload_s2_nbart.py:114]: how many places are there now that encode granule path construction? Maybe worth taking it out into a function?

Comment (Kirill888) [reply dags/sentinel_2_nbart/upload_s2_nbart.py:114]: Need to make sure that `upload_metadata` function produces the same paths as this checking code.

**Comment by kieranricardo** [comment]
Alrighty, I've changed a couple of things. The DAG now:
- uses the polygons from the old metadata if they exist and create them if they dont (20x speedup)
- skips already uploaded datsets
- change the date range from `prev_execution_date` - `execution_date` to `execution_date - 1 day` - `execution_date` (before the first dag run wouldn't do anything)

And I also checked that:
- `BETWEEN` is inclusive so we're catching all cases. I think this means we can get away with not using any overlap but maybe its still safe to overlap - whadya reckon @Kirill888?
- the AWS keys are only printed to the airflow logs, which still isn't great but at least its not on NCI

I also reduced the number of dags the path python```"s3://{S3_BUCKET}/{S3_PATH}/{granule_id}"``` path is constructed down from 4 to 2. There's a few other similar but not identical paths but I can't think of a good way to improve things further 🤷‍♂️

**Comment by Kirill888** [comment]
Paths are fine, I guess, we won't be changing those much.

Query, I'm not so sure. NCI does go down on occasion so assuming that this process will run every 24 hours might not be a good assumption.

**Comment by Kirill888** [comment]
I'm still not super clear how we are ensuring that we are getting ALL the new datasets

**Comment by Kirill888** [comment]
Specifically, what happens to datasets that were indexed after running this script but still on the same day as running this script. Are we including ALL the datasets that were added on the day of the last sync (some of which might have been synced already), or are we skipping ALL datasets that happened on the same day as last sync and hence might be loosing some datasets if someone is doing manual indexing late at night.

**Comment by Kirill888** [comment]
I think most sensible approach is to
1. Find all datasets added in `[last-sync-date 00:00, today 23:59]` time period
2. Skip those that are already on s3 (almost everything that was added on `last-sync-date` I assume, depending what time this runs)
3. Upload the new ones

**Comment by kieranricardo** [comment]
@Kirill888 good point!

I've changed it to now use `[prev_execution_date:00:00, execution_date + 1 day:00:00]`

**Review by Kirill888 — APPROVED** [review]

**Pull request merged**